File: problems/22-10-2025-S3=+-Stacks_and_Queues/Stacks/1047-remove-all-adjacent-duplicates-in-string/solution.py
# Time and Space Complexity are O(n)


class Solution:
    def removeDuplicates(self, s: str) -> str:
        """
        Idea is to recognize the LIFO pattern in the deletion order
        From there we know a stack is very useful here and we can simply iterate over and
        add elements into the stack and check if the top is the same as the current character
        """
        
        stack = []
        for c in s:
            if stack and stack[-1] == c:
                stack.pop()
            else:
                stack.append(c)
        
        return "".join(stack)

File: problems/22-10-2025-S3=+-Stacks_and_Queues/Stacks/1047-remove-all-adjacent-duplicates-in-string/solution.cpp
class Solution {
public:
    string removeDuplicates(string s) {
        // Strings in C++ are mutable, use the answer as a stack directly
        string ans = "";
        for (char c: s) {
            if (!ans.empty() && ans.back() == c) {
                ans.pop_back();
            } else {
                ans.push_back(c);
            }
        }
        
        return ans;
    }
};